### ISSUE TITLE
Dockerfile for people to quick-start contributing

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#32030]
It could be nice to have a Dockerfile with ccache and other things mentioned in CONTRIBUTING.md pre-configured and tweaked.

Also, the [README.md](https://raw.githubusercontent.com/pytorch/pytorch/master/README.md) forgets to mention CONTRIBUTING.md even in Contributing section (though this is not a big deal).

**Comment by ngimel** [comment]
cc @seemethere. #32515 addresses some of the issues.

**Comment by seemethere** [comment]
@vadimkantorov The root `Dockerfile` should now keep a consistent `ccache` between builds using [`buildkit`](https://github.com/moby/buildkit).

Usage is outlined in #32515 but I'll copy/paste it here:

## How to build:

Building the new images is pretty simple, just requires `docker` > 18.06 since the new build process relies on `buildkit` caching and multi-stage build resolving.

### Development images
For `runtime` images:
```
make -f docker.Makefile runtime-image
```

For `devel` images:
```
make -f docker.Makefile devel-image
```

Builds are tagged as follows:
```bash
docker.io/${docker_user:-whoami}/pytorch:$(git describe --tags)-${image_type}
```

Example:
```
docker.io/seemethere/pytorch:v1.4.0a0-2225-g9eba97b61d-runtime
```

**Comment by seemethere** [comment]
This is resolved by https://github.com/pytorch/pytorch/pull/32515

**Comment by vadimkantorov** [comment]
For development docker, it should probably accept as parameter the user’s fork URL, probably let the user build the code base themselves and let the user keep ccache on the host file system, ccache size should be configured to some large value, then instructions should be put in the main repo readme (along with expectations of build time).

The concern of this issue was to enable zeroconf dev environment for first time contributors. 

Cc @ngimel

**Comment by vadimkantorov** [comment]
Readme for “Building the image yourself” is not the same as “How to set up a dev/contributor environment using Docker”

**Comment by seemethere** [comment]
@vadimkantorov The base `Dockerfile` takes the current tree and copies it over to the docker build and does a fresh build from source with `ccache` that's implemented using the `buildkit` backend for docker.

I've outlined all of this in this PR: https://github.com/pytorch/pytorch/pull/32800/commits/0a0129b846c309e60d3e88f7503fecb437c9d4a8

It should be directly be a zero setup (aside from docker) build using:

```
make -f docker.Makefile devel-image
```

Do note that this build is only really for Linux as theres no actual way to do native macOS builds inside of Docker yet.

**Comment by vadimkantorov** [comment]
About Linux: sure. For other systems, Dockerfile could still be instructive as a sequence of preparing a dev env. Now this README is more clear! Thank you!

I mean the user story:
I use PyTorch for my own projects. I now want to develop and contribute a feature to PyTorch. I want this other PyTorch installation be separate from my main env. I don't want to be figuring out the reasons for super-long 1-hour long builds and want to have build caching set up by default (with good cache size). I want to have an instruction for rebuilding PyTorch (`setup.py install`? `setup.py develop`?) and cleaning the build artefacts if the build fails (sometimes I had to do this manually, for some reason `make clean` didn't work). Probably this instructions already exist but sprinkled all around the README.

**Comment by seemethere** [comment]
@vadimkantorov `buildkit` does not require any configuration for build size. I would recommend however clearing your build cache every now and then though with `docker builder prune`.

I believe the docker image approach should suffice for the user story you provided and I can understand the frustration involved with contributing to this project for the first time (I just started working on this project in December).

I think the user story you described is a bit different and more of a documentation issue rather than an issue about not having `docker` support in general.

I'll re-open this issue for now since we do not have sections in the `CONTRIBUTING.md` about the new docker build process in general.

**Comment by vadimkantorov** [comment]
Just checked, today still:
- no link to CONTRIBUTING.md from README.md
- almost no mention of Docker in CONTRIBUTING.md
- no clear guidance about fast builds / caching (ninja and ccache options are described with complex steps for ccache

Basically my setup is: I have working good and well conda-install of pytorch. I'd like to work on a PyTorch feature on the side with fast C++ builds. Is there a recommended script / sequence of steps I can execute that will create a conda env with all necessary dependencies and ccache/ninja config. Preferably it should also have a debug version of Python itself.

@ezyang I'd like to try to do some advances on inplace batchnorm :)

**Comment by ezyang** [comment]
@vadimkantorov If your distro is up to date enough you can just use the stock ccache; just setup the nvcc symlink. Do you want to Slack me and I can help you up? We don't have a dev setup script because everyone tends to have their own favorite distribution and generally we want people to be able to develop PyTorch on whatever system they work.

Generally debug build of Python isn't really necessary; I'm curious what you're hoping to use it for.

**Comment by vadimkantorov** [comment]
@ezyang Ok, I'll try to figure out ccache :) I'll ping you if it doesn't work :)

I think a conda dev script makes sense, since conda is the recommended install method for PyTorch, especially for new contributors who were only users of PyTorch and this may be their first big open-source project to contribute to (e.g. a lot of grad students who are not all professional devs in the sense of experience with toolchains for huge cuda/c++ projects fast building). Even if core devs have their own systems and don't need any dev script to kickstart, for new contributors it could be very useful.

**Comment by ezyang** [comment]
That's definitely reasonable. It's kind of hard to write a robust script that will work in all circumstances but it probably can be done with some elbow grease.

**Comment by vadimkantorov** [comment]
Maybe even a "recommended" annotated/commented compact sequence of comamnds with reasonable defaults for fast build times would be helpful. Parts of it are already in the testing Dockerfile, parts in Contrbiuting.md (super-gian btw:) ). It would be good to put in some variant  to README.

About debug Python builds: it helps if you have a crash dump. Then stack trace shows C++ functions, but nothing reasonable about Python - unless you have some gdb python extensions and maybe a debug version of python

**Comment by D1abloRUS** [comment]
@vadimkantorov maybe you can use https://github.com/puzl-ee/create-custom-docker-image

**Comment by vadimkantorov** [comment]
> @vadimkantorov maybe you can use https://github.com/puzl-ee/create-custom-docker-image

My point was mostly about recommended steps to set up a fast-building isolated environment, independent of the currently installed libraries/headers. On second thought, maybe this could be simply conda env instructions that include ccache (or whatever the modern version is)

When I finally get to doing PyTorch work, I may do this one too :) conda env instructions or Dockerfile

**Comment by veritas9872** [comment]
@vadimkantorov I have made a small project that may help with first-time contributors. Although I have not implemented the custom URL part yet, this would not be very difficult. Perhaps this project may be useful.
https://github.com/cresset-template/cresset
